### PR TITLE
[To Main] DESENG-656 - Add user management to engagement authoring wizard

### PR DESCRIPTION
**Pull request by NatSquared**
Issue #: [🎟️ DESENG-656](https://citz-gdx.atlassian.net/browse/DESENG-656)

~~⚠️ NOTE: PR #2555 should be merged first - this will decrease the number of changed files to 11~~

*Description of changes:*
- **Feature** Add user management to engagement authoring wizard
  - Added new UI component to handle searching, selecting, and displaying users
  - Minor tweaks to existing wizard functionality
    - Minor backend change: don't throw error when setting an engagement's slug to its current value

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-api/src/met_api/services/membership_service.py [hunk 1]
```diff
@@ -86,8 +86,8 @@ def _get_membership_details(user_details):
         default_role = CompositeRoles.TEAM_MEMBER.name
         default_membership_type = MembershipType.TEAM_MEMBER
 
-        is_reviewer = CompositeRoles.REVIEWER.value in user_details.get('main_role')
-        is_team_member = CompositeRoles.TEAM_MEMBER.value in user_details.get('main_role')
+        is_reviewer = user_details.get('main_role') == CompositeRoles.REVIEWER.value
+        is_team_member = user_details.get('main_role') == CompositeRoles.TEAM_MEMBER.value
```
Comment (NatSquared) [inline met-api/src/met_api/services/membership_service.py:90]: This was to fix a 500 error - main_role is a single string so it doesn't make much sense to ask if a role is "in" it

File: met-web/src/components/engagement/new/create/UserManager.tsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    useEffect(() => {
+        debouncedSearch.current(searchTerm);
+    }, [searchTerm]);
+
+    const users = (fetcher.data?.users?.items as User[] | undefined)?.filter(
+        (user) =>
+            // Remove ourselves from the list
+            user.id !== currentUser.userDetail.user?.id &&
+            // Remove admins from the list - they can already see everything and will cause an error if specified
+            user.main_role !== USER_COMPOSITE_ROLE.ADMIN.label,
```
Comment (Baelx) [inline met-web/src/components/engagement/new/create/UserManager.tsx:57]: Is this trailing comma necessary?

File: met-web/src/components/engagement/new/create/UserManager.tsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                        pr: 2,
+                        border: '1px solid',
+                        borderColor: 'primary.light',
+                        borderRadius: '8px',
+                        width: '300px',
+                    }}
+                >
+                    <Grid item>
+                        <BodyText bold color="primary.light">
+                            Team Member{selectedUsers.length > 1 && 's'} Added
```
Comment (Baelx) [inline met-web/src/components/engagement/new/create/UserManager.tsx:180]: Nice touch :). I probably would have just done "team member(s) added"

File: met-web/src/components/engagement/new/create/UserManager.tsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    const handleRemoveUser = (user: User) => {
+        setValue(
+            'users',
+            selectedUsers.filter((u) => u !== user),
+        );
+    };
+
+    return (
+        <Box width="100%">
+            <BodyText bold size="small">
```
Comment (Baelx) [inline met-web/src/components/engagement/new/create/UserManager.tsx:77]: Is this meant to be the label for the autocomplete field?

File: met-web/src/components/engagement/new/create/engagmentCreateAction.tsx [hunk 5]
```diff
@@ -22,6 +23,13 @@ export const engagementCreateAction: ActionFunction = async ({ request }) => {
     } catch (e) {
         console.error('Error updating engagement slug', e);
     }
+    formData
+        .get('users')
+        ?.toString()
+        .split(',')
```
Comment (Baelx) [inline met-web/src/components/engagement/new/create/engagmentCreateAction.tsx:29]: Is this converting an array to a string and back again? Is this a way to copy an array or something?

File: met-web/src/components/engagement/new/create/engagmentCreateAction.tsx [hunk 6]
```diff
@@ -22,6 +23,13 @@ export const engagementCreateAction: ActionFunction = async ({ request }) => {
...
+        .forEach(async (user_id) => {
```
Comment (Baelx) [inline met-web/src/components/engagement/new/create/engagmentCreateAction.tsx:30]: I don't think async/await works with forEach loops. https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

If you want to wait for the calls to finish before returning, you could use promise.all, for example

**Review by Baelx — CHANGES_REQUESTED** [review]
Looks good! Just one ask, I wanted to make sure the autocomplete field was labelled in some way. Let me know if I missed it.

**Review by Baelx — APPROVED** [review]
Thanks for the label!

File: met-web/src/components/engagement/new/create/UserManager.tsx (hunk 2 above)
Comment (Baelx) [reply met-web/src/components/engagement/new/create/UserManager.tsx:57]: 😵‍💫

**Pull request merged**